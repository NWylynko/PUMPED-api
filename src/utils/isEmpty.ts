export const isArrayEmpty = (array: any[]): true | undefined => (array.length === 0 ? true : undefined);

export const isEmpty = (object: any): true | undefined => (!object ? true : undefined);
