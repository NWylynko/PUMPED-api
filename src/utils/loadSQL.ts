import fs from 'fs';
import path from 'path';
import sqlParser from 'sql-template-tag';

const loadSQL = (file: string) => {
  const sql = fs.readFileSync(path.resolve(__dirname, '../../SQL/', `${file}.SQL`), 'utf-8');

  return sqlParser([sql]).sql;
};

export default loadSQL;
