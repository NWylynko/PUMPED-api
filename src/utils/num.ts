const num = (n: string) => parseInt(n, 10);

export default num;
