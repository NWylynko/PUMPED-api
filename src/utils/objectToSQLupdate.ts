import { escape } from 'sqlstring';

const objectToSQLupdate = (fields: any) => {
  let sql = '';

  Object.keys(fields).forEach((partName, index, array) => {
    // if the current item is the length of the array then its the last item
    // the last item cant have a comma
    const comma = index === array.length - 1 ? '' : ',';

    sql += `${escape(partName)} = ${escape(fields[partName])}${comma} `;
  });

  return sql;
};

export default objectToSQLupdate;
