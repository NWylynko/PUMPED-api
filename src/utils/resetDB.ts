import { promises as fs } from 'fs';
import db, { SQL } from '../db';

export const createTables = () => Promise.all([
  db.run(SQL.tables.Brand),
  db.run(SQL.tables.Collection),
  db.run(SQL.tables.Colour),
  db.run(SQL.tables.ColourImage),
  db.run(SQL.tables.Customer),
  db.run(SQL.tables.Image),
  db.run(SQL.tables.Order),
  db.run(SQL.tables.OrderItem),
  db.run(SQL.tables.Review),
  db.run(SQL.tables.Section),
  db.run(SQL.tables.Shoe),
  db.run(SQL.tables.ShoeTag),
  db.run(SQL.tables.Stock),
  db.run(SQL.tables.Style),
  db.run(SQL.tables.Tag),
  db.run(SQL.tables.WishList),
]);

const resetDB = async () => {
  await db.close();
  await fs.unlink(`./main.${process.env.NODE_ENV}.db`);
  await db.open();
  await createTables();
};

export const addTestData = () => db.exec(SQL.testData);

export const resetDBForTest = async () => {
  await db.close();
  await db.open();
  await createTables();
  await db.exec(SQL.testData);
};

export default resetDB;
