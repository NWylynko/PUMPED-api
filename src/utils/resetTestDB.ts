import { promises as fs } from 'fs';
import db from '../db';

const reset = async () => {
  await db.close();
  await fs.unlink('./main.test.db');
  await fs.copyFile('./main.testData.db', './main.test.db');
  await db.open();
};

export default reset;
