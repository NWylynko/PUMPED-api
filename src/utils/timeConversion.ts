function timeConversion(milliseconds: number) {
  // adapted from https://stackoverflow.com/a/32180863

  const seconds = (milliseconds / 1000);
  const minutes = (milliseconds / (1000 * 60));
  const hours = (milliseconds / (1000 * 60 * 60));
  const days = (milliseconds / (1000 * 60 * 60 * 24));

  if (seconds < 60) {
    return `${Math.floor(seconds)} Seconds`;
  } if (minutes < 60) {
    return `${Math.floor(minutes)} Minutes`;
  } if (hours < 24) {
    return `${Math.floor(hours)} Hours`;
  }
  return `${Math.floor(days)} Days`;
}

export default timeConversion;
