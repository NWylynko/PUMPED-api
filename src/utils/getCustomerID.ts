import { Request } from 'express';
import num from './num';

const getCustomerID = (req: Request): number => {
  const CustomerID = req.header('CustomerID');

  if (!CustomerID) {
    throw new Error('no customer id supplied');
  }

  return num(CustomerID);
};

export default getCustomerID;
