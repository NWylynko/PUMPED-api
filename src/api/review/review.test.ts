// eslint-disable-next-line import/no-extraneous-dependencies
import supertest from 'supertest';
import express from '../../app';
import { resetDBForTest } from '../../utils/resetDB';

const app = supertest(express);

beforeEach(async (done) => { await resetDBForTest(); done(); });

describe('GET Review of Shoe', () => {
  [1, 2, 3].forEach(async (ShoeID) => {
    it(`gets reviews for ShoeID (${ShoeID})`, async () => {
      const response = await app.get(encodeURI(`/review/${ShoeID}`));

      expect(response.status).toMatchSnapshot();

      const json = JSON.parse(response.text);

      expect(json).toMatchSnapshot();
    });
  });
});

describe('POST Review of Shoe', () => {
  it('add review with no data (should error)', async () => {
    const review = {};

    const addReview = await app
      .post('/review/1/1')
      .type('json')
      .send(JSON.stringify(review));

    expect(addReview.status).toMatchSnapshot();

    const addReviewJson = JSON.parse(addReview.text);

    expect({ error: addReviewJson.error, message: addReviewJson.message }).toMatchSnapshot();

    const getReview = await app
      .get('/review/1');

    expect(getReview.status).toMatchSnapshot();

    const getReviewJson = JSON.parse(getReview.text);

    expect(getReviewJson).toMatchSnapshot();
  });

  it('add review from customer who already left a review (should error)', async () => {
    const review = {
      stars: 2,
      message: 'baddd',
      timestamp: 309458,
    };

    const addReview = await app
      .post('/review/1/1')
      .type('json')
      .send(JSON.stringify(review));

    expect(addReview.status).toMatchSnapshot();

    const addReviewJson = JSON.parse(addReview.text);

    expect(addReviewJson).toMatchSnapshot();

    const getReview = await app
      .get('/review/1');

    expect(getReview.status).toMatchSnapshot();

    const getReviewJson = JSON.parse(getReview.text);

    expect(getReviewJson).toMatchSnapshot();
  });

  it('add review (should work)', async () => {
    const review = {
      stars: 4,
      message: 'nice',
      timestamp: 3456745367,
    };

    const addReview = await app
      .post('/review/3/3')
      .type('json')
      .send(JSON.stringify(review));

    expect(addReview.status).toMatchSnapshot();

    const addReviewJson = JSON.parse(addReview.text);

    expect(addReviewJson).toMatchSnapshot();

    const getReview = await app
      .get('/review/3');

    expect(getReview.status).toMatchSnapshot();

    const getReviewJson = JSON.parse(getReview.text);

    expect(getReviewJson).toMatchSnapshot();
  });
});

describe('PATCH (update) Review of Shoe', () => {
  it('update review with no data (should error)', async () => {
    const reviewDetails = {};

    const updateReview = await app
      .patch('/review/1/1')
      .type('json')
      .send(JSON.stringify(reviewDetails));

    expect(updateReview.status).toMatchSnapshot();

    const updateReviewJson = JSON.parse(updateReview.text);

    expect({ error: updateReviewJson.error, message: updateReviewJson.message }).toMatchSnapshot();

    const getReview = await app
      .get('/review/1');

    expect(getReview.status).toMatchSnapshot();

    const getReviewJson = JSON.parse(getReview.text);

    expect(getReviewJson).toMatchSnapshot();
  });

  it("update review that doesn't exist (should error)", async () => {
    const reviewDetails = {
      stars: 2,
      message: 'baddd',
      timestamp: 309458,
    };

    const updateReview = await app
      .patch('/review/1/8')
      .type('json')
      .send(JSON.stringify(reviewDetails));

    expect(updateReview.status).toMatchSnapshot();

    const updateReviewJson = JSON.parse(updateReview.text);

    expect(updateReviewJson).toMatchSnapshot();

    const getReview = await app
      .get('/review/1');

    expect(getReview.status).toMatchSnapshot();

    const getReviewJson = JSON.parse(getReview.text);

    expect(getReviewJson).toMatchSnapshot();
  });

  it('add review (should work)', async () => {
    const reviewDetails = {
      stars: 4,
      message: 'nice',
      timestamp: 3456745367,
    };

    const addReview = await app
      .patch('/review/3/1')
      .type('json')
      .send(JSON.stringify(reviewDetails));

    expect(addReview.status).toMatchSnapshot();

    const addReviewJson = JSON.parse(addReview.text);

    expect(addReviewJson).toMatchSnapshot();

    const getReview = await app
      .get('/review/3');

    expect(getReview.status).toMatchSnapshot();

    const getReviewJson = JSON.parse(getReview.text);

    expect(getReviewJson).toMatchSnapshot();
  });
});

describe('DELETE (remove) Review of Shoe', () => {
  it('remove review (should work)', async () => {
    const removeReview = await app
      .delete('/review/1/1');

    expect(removeReview.status).toMatchSnapshot();

    const removeReviewJson = JSON.parse(removeReview.text);

    expect(removeReviewJson).toMatchSnapshot();

    const getReview = await app
      .get('/review/1');

    expect(getReview.status).toMatchSnapshot();

    const getReviewJson = JSON.parse(getReview.text);

    expect(getReviewJson).toMatchSnapshot();
  });
});
