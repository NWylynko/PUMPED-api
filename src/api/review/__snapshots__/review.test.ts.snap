// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DELETE (remove) Review of Shoe remove review (should work) 1`] = `200`;

exports[`DELETE (remove) Review of Shoe remove review (should work) 2`] = `
Object {
  "data": Object {
    "CustomerID": "1",
    "ShoeID": "1",
  },
  "success": true,
}
`;

exports[`DELETE (remove) Review of Shoe remove review (should work) 3`] = `200`;

exports[`DELETE (remove) Review of Shoe remove review (should work) 4`] = `
Object {
  "data": Array [
    Object {
      "firstName": "pippy",
      "lastName": "nippy",
      "message": "great!",
      "stars": 4,
      "timestamp": 20938475,
    },
  ],
  "success": true,
}
`;

exports[`GET Review of Shoe gets reviews for ShoeID (1) 1`] = `200`;

exports[`GET Review of Shoe gets reviews for ShoeID (1) 2`] = `
Object {
  "data": Array [
    Object {
      "firstName": "bobby",
      "lastName": "boo",
      "message": "good",
      "stars": 3,
      "timestamp": 35687,
    },
    Object {
      "firstName": "pippy",
      "lastName": "nippy",
      "message": "great!",
      "stars": 4,
      "timestamp": 20938475,
    },
  ],
  "success": true,
}
`;

exports[`GET Review of Shoe gets reviews for ShoeID (2) 1`] = `200`;

exports[`GET Review of Shoe gets reviews for ShoeID (2) 2`] = `
Object {
  "data": Array [
    Object {
      "firstName": "bobby",
      "lastName": "boo",
      "message": "terrable!!",
      "stars": 2,
      "timestamp": 2984577673,
    },
  ],
  "success": true,
}
`;

exports[`GET Review of Shoe gets reviews for ShoeID (3) 1`] = `200`;

exports[`GET Review of Shoe gets reviews for ShoeID (3) 2`] = `
Object {
  "data": Array [],
  "empty": true,
  "success": true,
}
`;

exports[`PATCH (update) Review of Shoe add review (should work) 1`] = `200`;

exports[`PATCH (update) Review of Shoe add review (should work) 2`] = `
Object {
  "data": Object {
    "CustomerID": "3",
    "ShoeID": "1",
    "message": "nice",
    "stars": 4,
    "timestamp": 3456745367,
  },
  "success": true,
}
`;

exports[`PATCH (update) Review of Shoe add review (should work) 3`] = `200`;

exports[`PATCH (update) Review of Shoe add review (should work) 4`] = `
Object {
  "data": Array [],
  "empty": true,
  "success": true,
}
`;

exports[`PATCH (update) Review of Shoe update review that doesn't exist (should error) 1`] = `200`;

exports[`PATCH (update) Review of Shoe update review that doesn't exist (should error) 2`] = `
Object {
  "data": Object {
    "CustomerID": "1",
    "ShoeID": "8",
    "message": "baddd",
    "stars": 2,
    "timestamp": 309458,
  },
  "success": true,
}
`;

exports[`PATCH (update) Review of Shoe update review that doesn't exist (should error) 3`] = `200`;

exports[`PATCH (update) Review of Shoe update review that doesn't exist (should error) 4`] = `
Object {
  "data": Array [
    Object {
      "firstName": "bobby",
      "lastName": "boo",
      "message": "good",
      "stars": 3,
      "timestamp": 35687,
    },
    Object {
      "firstName": "pippy",
      "lastName": "nippy",
      "message": "great!",
      "stars": 4,
      "timestamp": 20938475,
    },
  ],
  "success": true,
}
`;

exports[`PATCH (update) Review of Shoe update review with no data (should error) 1`] = `500`;

exports[`PATCH (update) Review of Shoe update review with no data (should error) 2`] = `
Object {
  "error": true,
  "message": "no data sent",
}
`;

exports[`PATCH (update) Review of Shoe update review with no data (should error) 3`] = `200`;

exports[`PATCH (update) Review of Shoe update review with no data (should error) 4`] = `
Object {
  "data": Array [
    Object {
      "firstName": "bobby",
      "lastName": "boo",
      "message": "good",
      "stars": 3,
      "timestamp": 35687,
    },
    Object {
      "firstName": "pippy",
      "lastName": "nippy",
      "message": "great!",
      "stars": 4,
      "timestamp": 20938475,
    },
  ],
  "success": true,
}
`;

exports[`POST Review of Shoe add review (should work) 1`] = `200`;

exports[`POST Review of Shoe add review (should work) 2`] = `
Object {
  "data": Object {
    "CustomerID": "3",
    "ShoeID": "3",
    "message": "nice",
    "stars": 4,
    "timestamp": 3456745367,
  },
  "success": true,
}
`;

exports[`POST Review of Shoe add review (should work) 3`] = `200`;

exports[`POST Review of Shoe add review (should work) 4`] = `
Object {
  "data": Array [
    Object {
      "firstName": "whip",
      "lastName": "nah",
      "message": "nice",
      "stars": 4,
      "timestamp": 3456745367,
    },
  ],
  "success": true,
}
`;

exports[`POST Review of Shoe add review from customer who already left a review (should error) 1`] = `500`;

exports[`POST Review of Shoe add review from customer who already left a review (should error) 2`] = `
Object {
  "error": true,
  "message": "SQLITE_CONSTRAINT",
}
`;

exports[`POST Review of Shoe add review from customer who already left a review (should error) 3`] = `200`;

exports[`POST Review of Shoe add review from customer who already left a review (should error) 4`] = `
Object {
  "data": Array [
    Object {
      "firstName": "bobby",
      "lastName": "boo",
      "message": "good",
      "stars": 3,
      "timestamp": 35687,
    },
    Object {
      "firstName": "pippy",
      "lastName": "nippy",
      "message": "great!",
      "stars": 4,
      "timestamp": 20938475,
    },
  ],
  "success": true,
}
`;

exports[`POST Review of Shoe add review with no data (should error) 1`] = `500`;

exports[`POST Review of Shoe add review with no data (should error) 2`] = `
Object {
  "error": true,
  "message": "no data sent",
}
`;

exports[`POST Review of Shoe add review with no data (should error) 3`] = `200`;

exports[`POST Review of Shoe add review with no data (should error) 4`] = `
Object {
  "data": Array [
    Object {
      "firstName": "bobby",
      "lastName": "boo",
      "message": "good",
      "stars": 3,
      "timestamp": 35687,
    },
    Object {
      "firstName": "pippy",
      "lastName": "nippy",
      "message": "great!",
      "stars": 4,
      "timestamp": 20938475,
    },
  ],
  "success": true,
}
`;
